### ISSUE TITLE
Report detailed error message about why node save failed

### ISSUE THREAD

**Issue by lassoan** [Slicer/Slicer#4968]
Save data dialog may show several popups when nodes cannot be saved. The popups only tell that saving failed, but not why or what the user can do about it.

We should show a single popup and/or show comments (or error icons) in the save data dialog table. Error message should be more specific.

See for example these related issues:
https://mantisarchive.slicer.org/view.php?id=4714
https://github.com/Slicer/SlicerGitSVNArchive/commit/d596340895ed6e54bc164a6715b3b030654129ee

**Comment by lassoan** [comment]
It should not be hard to improve this:
1. we could easily replace repetitive popups with a single popup at the end; adding a new column for status text or icon would not be hard either
2. we could store a LastErrorDescription member variable in MRML storage nodes, which could contain a more specific, human-readable error message

**Comment by pieper** [comment]
Yes, and for loading too.

**Comment by Leengit** [comment]
For those who weren't at the 6/9/20 hangout, the idea is to modify storage nodes to collect warnings and errors and then to present them nicely in a summary dialog box for the user ... not just outputting developer-style text to the console.  I'd like to sanity check these possibilities:

1. We could replace every invocation of the vtkWarningMacro in every Slicer/Libs/MRML/Core/vtkMRML*StorageNode.cxx file with a new macro storageNodeWarningMacro.  The new macro would be in charge of (1) logging the warning in a user-friendly way, perhaps via a new (overridable) vtkMRMLStorageNode::logWarningMessage method, and (2) passing the arguments to vtkWarningMacro for traditional handling.  At opportune places near the end of some vtkMRML*StorageNode methods we'd call a new vtkMRMLStorageNode::showWarningMessage method, to summarize the collected information into a single user-friendly dialog box.  (And likewise with "Warning" replaced with "Error" in all of the above.)
2.     I see that vtkWarningWithObjectMacro (which is invoked by vtkWarningMacro) checks whether the object->HasObserver("WarningEvent") and, if so, invokes the event.  Instead of idea #1, could I add some sort of an observer that collects and creates a user-friendly version of the messages?  Although this collects all the messages, I don't know how a post-processing that displays the summary in a single dialog box would get invoked -- perhaps much like idea #1, we could somehow invoke the observer in some distinct way, at opportune places in the vtkMRML*StorageNode methods.
3.     Maybe, I am significantly off the mark and there is a better approach?  If so, what am I not considering?

Thanks!

**Comment by lassoan** [comment]
We don't need to mess with the warning macros, they log detailed information (such as filename, line number, etc) that is useful for developers. For users, we need different kind of messages, sometimes usually higher level, and potentially explaining what the user can do about the error. Ideally, they should also be translatable, but since translation happens in GUI classes, we can keep it fixed to English now.

I think all we need are the followings:

1. New WarningMessages and ErrorMessages members in vtkMRMLStorageNode, with type of vtkStringArray (as it can be made accessible in Python directly - as opposed to std::vector<std::string>, which I think is still not Python-wrapped). The advantage of string list as opposed to a simple string is that we can count the number of errors and we have flexibility in how we format the messages on the GUI (we can display them in a HTML list, or in a table, etc.).

2. Add warning/error messages to the message lists. All places where we write messages to the log are good candidates for displaying messages to users, too.

3. When reading/writing is completed, the GUI would check if the storage node has any errors and warnings and would display them in a popup. Even something simple like this could be OK:

    slicer.util.confirmRetryCloseDisplay("15 errors occurred while writing files a.nrrd, n.nrrd, c.nrrd. ", detailedText="- a.nrrd: cannot write folder\n- n.nrrd: empty image")

**Comment by pieper** [comment]
Glad I didn't start typing before @lassoan's comment appears.  I agree with his points.

An architecture heads up @Leengit, we wouldn't have a `vtkMRMLStorageNode::showWarningMessage` because MRML classes are never linked with GUI code in Slicer.  MRML nodes are primarily data holders, with some small helper methods to manipulate the data.  That's why we prefer the warning/error message lists that can be acted on in the application gui code, sometimes with events like your option number two.

**Comment by jcfr** [comment]
> `std::vector<std::string>`, which I think is still not Python-wrapped)

These are now supported. See https://github.com/Kitware/VTK/blob/master/Wrapping/Python/README.md#stl-containers

> For users, we need different kind of messages, sometimes usually higher level, and potentially explaining what the user can do about the error.

:+1: 

> the GUI would check if the storage node has any errors

To capture warnings and errors, I suggest to use https://github.com/Slicer/vtkAddon/blob/master/vtkErrorSink.h

**Comment by lassoan** [comment]
>>std::vector<std::string>, which I think is still not Python-wrapped)
>
> These are now supported. 

Do you mean in VTK 9? (I don't think these work in Slicer's current VTK version)
This will be great! We will be able to remove some methods that we added for Python compatibility and we'll get much more convenient interface in Python.

**Comment by Leengit** [comment]
As many of you recommended at the Slicer Hangout today, I won't be forming the user-friendly version of warnings and errors at this time.  Instead, we will be recording the developer-friendly warnings and errors with a vtkErrorSink object allocated by and observing an object with a  vtkMRMLStorageNode base class.  The existing invocations of vtkErrorMacro and vtkWarningMacro will then automatically log to the vtkErrorSink object.  This is very easy to instantiate on the recording side, with vtkErrorSink::New and vtkErrorSink::SetObservedObject.

I need help as to where these recorded messages should be detected and relayed to the user.  Would that be best in vtkMRMLScene::AddNode?

**Comment by lassoan** [comment]
Collect error and warning messages from storage nodes and then display them to the user in the Save data/Add data Qt widgets. Some error reporting is already done during saving data, it is just not convenient (error popup is shown for each message instead of showing a single window with a summary of of all issues; only errors can be reported, not warnings; there is no easy way of seeing more details - user is asked to check the logs, but there are lots of information in the logs), so it could be easier to start with the save dialog.

**Comment by Leengit** [comment]
My grep'ing for keywords that I think would be applicable is not turning up the right files I think.  What file(s) contain(s) the "Save data/Add data Qt widgets"?  Something in Libs/MRML/Widgets or QTCore or QTGUI or QTApp or ...?

**Comment by jcfr** [comment]
Reading this page will be helpful to understanding the bigger picture. See https://www.slicer.org/wiki/Documentation/Nightly/Developers/IO

**Comment by pieper** [comment]
Yes, I think the end-user experience that Andras describes makes sense, and starting with the save dialog is good.

@Leengit , in addition to the reference posted by Jc, if you haven't already it would make sense to step through the save dialog with a debugger to see all the layers involved between VTK/ITK writers, MRML storage nodes, and Qt widgets and dialogs.  Setting a breakpoint in the writer will show you the whole stack trace so you can find the corresponding files.  There are several architectural layers that are intentionally isolated so you need to look at existing code for models about how data is passed.

**Comment by lassoan** [comment]
This is the message box to be replaced: https://github.com/Slicer/Slicer/blob/43300c7b9a3b27265c2bc736693cb881f05e3ea9/Base/QTGUI/qSlicerSaveDataDialog.cxx#L886

You could display a summary popup in the same method, outside the for loop:

![image](https://user-images.githubusercontent.com/307929/84840663-36953400-b00e-11ea-9520-3f8c9d151e7c.png)

However, this implementation would not be ideal, as users would need to review and remember all warning/error messages and which nodes they were for, if they want to fix it (by changing file location, file format, etc.). Therefore, I think it would be much better to add one (or a few) new columns to the save data dialog that could show the saving results:

![image](https://user-images.githubusercontent.com/307929/84840990-19149a00-b00f-11ea-8c7c-76d70fb4b312.png)

We could show only the number of warning/error messages only and clicking "Details" button could show all the original low-level messages in a popup. A checkbox could be added to only show nodes that had saving errors/warnings to make it easier to see what went wrong. Users could make adjustments and click Save again. The window is closed after user clicks "Save" if there are no warnings or errors during saving of the selected nodes (or user clicks "Cancel").

**Comment by Leengit** [comment]
After some detours I am back to working on this issue.  It appears that there are a number of node types that get saved by `qSlicerSaveDataDialogPrivate::saveNodes` but I have added `vtkErrorSink` technology to only the `vtkMRMLStorageNode` class.  So, for errors and warnings that are generated within `qSlicerSaveDataDialogPrivate::saveNodes` itself (such as "file name is empty"), should I have them continue to be one-off pop-up dialogs or should I instead log them via the `vtkErrorSink` in the case of a `vtkMRMLStorageNode` (henceforth always including its subclasses)?

And what if the call to `coreIOManager->saveNodes(fileType, savingParameters)` returns false?  I see no alternative to continuing to do a one-off dialog if the node is not a `vtkMRMLStorageNode`.  In the case that the node is `vtkMRMLStorageNode`, I am thinking that I do not have a pop up.  Nor do I log an additional error if the `vtkErrorSink` of the `vtkMRMLStorageNode` has any messages already.

I still need to tackle how we will display, in the save data dialog or independent pop-up, any recorded errors and warnings from the `vtkErrorSink` objects of the `vtkMRMLStorageNode` objects.  That some of the saved objects may not be `vtkMRMLStorageNode` objects could complicate that.

**Comment by pieper** [comment]
I think it would be best to use the design where any error reporting is integrated in the Save Data dialog, so that the user can fix any identified issues, like by picking a different filename to avoid overwriting existing data or pick a different directory if the path is read-only.  Any nodes that had no error on write would get unchecked in the first column and maybe the ones that failed get moved to the top of the list.

Also, a lot of the logic in `qSlicerSaveDataDialogPrivate::saveNodes` could be added to a validator class that gets called on every event in the dialog so that the Save button is only enabled when, for example, there are no empty filenames and all the paths are writable.  That would also be a good place to add flags if any of the selected save formats are lossy for the selected data (could add a virtual method like `vtkMRMLStorageNode::IsLossy` which would return true or false for the current modes and storable node).

In terms of MRML, only `vtkMRMLStorable` subclasses are stored, so the `vtkMRMLStorageNode` is the correct place to focus the error sink checking.

**Comment by lassoan** [comment]
> I think it would be best to use the design where any error reporting is integrated in the Save Data dialog, so that the user can fix any identified issues, like by picking a different filename to avoid overwriting existing data or pick a different directory if the path is read-only. 

Fully agree. See the mockup above: https://github.com/Slicer/Slicer/issues/4968#issuecomment-645077698

> In terms of MRML, only vtkMRMLStorable subclasses are stored, so the vtkMRMLStorageNode is the correct place to focus the error sink checking.

Exactly. For now, you can implement just a few checks in vtkMRMLStorageNode. You don't even need to bother with vtkErrorSink, just do some trivial checks as a proof of concept. We'll add more checks later.

**Comment by Leengit** [comment]
> In terms of MRML, only vtkMRMLStorable subclasses are stored, so the vtkMRMLStorageNode is the correct place to focus the error sink checking.

Yes, that is where I am focused.  However, I am seeing objects that are not derived from `vtkMRMLStorageNode`, which can generate errors within `qSlicerSaveDataDialogPrivate::saveNodes`, such as file name is empty and `coreIOManager->saveNodes(fileType, savingParameters)` returns false.  I am seeing a `vtkMRMLScalarVolumeNode` which is `vtkMRMLStorableNode` but not a `vtkMRMLStorageNode` and perhaps there are others?  Are you saying that the non-`vtkMRMLStorageNode` cases should be left as one-off dialog boxes for now or that they should be addressed at present ... or that they are not relevant for some reason?

> Exactly. For now, you can implement just a few checks in vtkMRMLStorageNode. You don't even need to bother with vtkErrorSink, just do some trivial checks as a proof of concept. We'll add more checks later.

By using `vtkErrorSink`, all the existing invocations of `vtkWarningMacro` and `vtkErrorMacro` currently present in `vtkMRMLStorageNode` methods will write to the `vtkErrorSink`.  Are you saying that instead we use separate `vtkStringArray` (or `std::vector<std::string>`) objects for warnings and errors and add code adjacent to the macro invocations to append to these arrays?

> I think it would be best to use the design where any error reporting is integrated in the Save Data dialog, so that the user can fix any identified issues, like by picking a different filename to avoid overwriting existing data or pick a different directory if the path is read-only.

If I understand correctly, that means that instead of displaying anything from `qSlicerSaveDataDialogPrivate::saveNodes`, I should instead put these warnings or errors into a `vtkErrorSink` object (or into a pair of `vtkStringArray` objects) that would (somehow) be accessed by the Save Data dialog.  Is that right?  Does the answer depend upon whether the warning/error originating from `qSlicerSaveDataDialogPrivate::saveNodes` is for a `vtkMRMLStorageNode` object or something else like a `vtkMRMLScalarVolumeNode` object?

I am thinking that it is *not* the `qSlicerSaveDataDialogPrivate::save` method (or anything that it calls) that would be displaying the Save Data dialog (as the mockup above).  If you can point me in the right direction for the method that throws up the main Save Data dialog that would help ... I am having trouble navigating the GUI files.

**Comment by pieper** [comment]
I empathize; the system is complex and takes some time to navigate.  But the layers exist for good reasons and we want to maintain the independent and reusable nature of these layers.  It might help to set breakpoints in the storage nodes and look at the call tree to see what is calling what.

In MRML, being a [storable node](https://apidocs.slicer.org/master/classvtkMRMLStorableNode.html) means that you have external bulk storage (one or more files), which is different from a generic node which serializes data to the MRML XML.  The [storage node](https://apidocs.slicer.org/master/classvtkMRMLStorageNode.html) is what implements the read/write.  So yes, that's the right place to catch errors at the data level, because it's what the `coreIOManager` uses under the hood.  

These storage nodes don't know about the GUI, so we need to collect any errors and report them to the user.  Some errors are caught at the storage node level made available to the gui to report back to the user, as Andras says.  Yes, string arrays of error and warning messages would be a good way to do this.  Some of the writers use third party libraries that may only report issues to stderr, which is why the `vtkErrorSink` can be useful to populate the string array, but we can start with issues detectable at the storage node level.  

So basically each layer should have a way to report warning and error messages up the chain, from libraries, to storage nodes, to the coreIOManager, to the widget.  These should be in the form of success/fail return codes and introspectable state that can be inspected by the user like in the mockup above.

The other types of errors are widget level things, like having an empty filename in the dialog, and those are under our control, which is why the error dialogs already exist, but the user experience could be cleaner if all similar types of errors were handled consistently.  (or prevented, by not enabling the save button until the file names and paths are all valid).

**Comment by lassoan** [comment]
I think @Leengit is considering using reporting errors from vtkMRMLStorageNode to the GUI by emitting warning/error events. Progress and error reporting mechanism in Slicer is quite basic now (we mostly just rely on user to check the log if the icon is highlighted), so there is lots of room for improvement, and an event-based mechanism for detecting and propagating errors could work well. However, this should be a larger redesign that considers all use cases (data reading, writing, and foreground and background processing). That's why we recommend to go with the simple solution that we discussed (store error/warnings of the last operation in the storage node).

**Comment by Leengit** [comment]
Since we will likely want to record warning and error messages for other than `vtkMRMLStorageNode` objects, I am thinking that we should implement this as a separate class and in separate files from `vtkMRMLStorageNode`.  The new class would be like `vtkErrorSink` in that it manages messages, but different in that it won't be observing other objects, etc.

What should the new class be ***called*** and what ***directory*** should its implementation files be added to?  For example:

- `Slicer/Libs/MRML/Core/vtkMessageLog.h` -- which is the same directory where `vtkMRMLStorageNode` lives
- `VTK/Common/Core/vtkErrorLog.h` -- which is the same directory where `vtkCommand` lives
- `vtkAddon/vtkMessageRecords.h`-- which is the same directory where `vtkErrorSink` lives

... or some mix and match of the above projects/directories and class/file names.  Internally, the messages will probably be stored much as they are in `vtkErrorSink`, which is as a `std::vector<vtkErrorSink::Message>`.

**Comment by pieper** [comment]
Yes, it's true that other operations besides storage could generate error and warning messages, but these are already in the error log.  To me the Save Data dialog is the worst offender and a reasonable scope to address, and the storage node is the right place to capture the info.

**Comment by lassoan** [comment]
I would recommend to go with the simplest design - add a few member variables and methods in vtkMRMLStorageNode. Later, when we are ready to improve error reporting for other nodes (e.g., CLI module nodes for data processing) then we can refactor/generalize the design more. This would allow us to focus on the problem we want to solve now (improve user feedback during data saving) and make the solution available in the upcoming Slicer5 release.

**Comment by Leengit** [comment]
At the moment I am displaying "Message: All good!" for debug purposes.  Normally that will be blank unless there are actual user-style Warning or Error messages.  At present Warning and Error messages change that text to something like
```
Warning: Could not find my navel.
Error: Toe jam football.
```

Assuming that this is on the right track, having actual warning and error icons and other prettier formatting would be a next step.

![Screenshot from 2020-07-24 10-50-58](https://user-images.githubusercontent.com/35611400/88404319-bba50300-cd9b-11ea-85f9-4d408ef42a75.png)

**Comment by pieper** [comment]
I got an email from github with a screenshot, but it's not here now.  In any case it looked good and my only suggestion would be that the User Messages column only appear if there is an error or warning to display.

**Comment by Leengit** [comment]
The [screenshot](https://user-images.githubusercontent.com/35611400/88404319-bba50300-cd9b-11ea-85f9-4d408ef42a75.png) is over at [Pull 5045](https://github.com/Slicer/Slicer/pull/5045).

**Comment by jcfr** [comment]
Closing. Fixed in https://github.com/Slicer/Slicer/commit/517ded4fd4fb6e69c48b042bc7ade36fadf0f29f associated with https://github.com/Slicer/Slicer/pull/5045
